### DISCUSSION TITLE
Question: Does unpoly support merge of the head content?

### DISCUSSION THREAD

**Discussion by didacusAbella** [unpoly/unpoly#296] [Q&A] [answered]
I used unpoly and i noticed when I switch between pages if the new page has new content in the head it isn't merged with the old. Does unpoly support this feature by some config or api?

**Comment by adam12** [comment, marked as answer]
Unpoly will merge `<title>` from new pages. The rest of `<head>` is ignored, AFAIK.

**Reply by didacusAbella** [reply]
Thanks! Really a shame. Do you know if there is some sort of lifecycle event that I can intercept before unpoly swap the content? I want to try write some custom code in my project to implement this feature.

**Reply by adam12** [reply]
Maybe you can listen for [`up:fragment:loaded`](https://unpoly.com/up:fragment:loaded).

What changed in `<head>` that you need per-request?

**Reply by didacusAbella** [reply]
I want to inject some specific page styles or scripts. For example I have only one page that use Chart.js for display some stats and I want to load it on demand.

**Reply by adam12** [reply]
`<link>` tags can come with the response. They don't need to live in `<head>`.

For `<script>` tags, Unpoly won't parse them by default, but there's a better way to load on-demand. See this [Makandra card](https://makandracards.com/makandra/52361-unpoly-loading-large-libraries-on-demand).

**Reply by Nefiron** [reply]
I asked on twitter about this and apparantly you can use [up-hungry] on the meta tags so they get replaced if necessary. :)

**Reply by janreges** [reply]
Hi @adam12. We love libraries like Unpoly/HTMX/Turbo and huge disadvantage of Unpoly is non functional <head> merge.

For example, when you use Astro for building websites, build process automatically generates and links extra CSS/JS to <head> for each page separatelly, so each page loading only really necessary CSS/JS.

Turbo has very well working <head> merging and HTMX will contains <head> merging in next version 1.8.3. We love Unpoly for many exceptional reasons, but the inability to merge <head> with CSS/JS is a major drawback. It would be great if this could be implemented. I think that the implementation itself could be inspired by other libraries.

What do you think?

**Reply by adam12** [reply]
@janreges I wouldn't be the one making the call to add this to Unpoly (it would be @triskweline). I'm reasonably neutral on the feature itself.

**Reply by triskweline** [reply]
I'm interested in solving this, in particular to better support splitting large asset bundles without a build tool that can do chunks (like Webpack). However if Unpoly is to have a stronger opinion over how assets (JavaScripts and stylesheets) are delivered, it must either have a solid design or offer a lot of configuration to fit everyone's use case.

Some of decisions that need to be made:

- Whether assets are per-screen or per-component
- Whether assets are included in the `<head>` or closer to the elements they're modifying
- Whether rendering waits for assets to be loaded (to support compiler priorities and prevent flashes of uncompiled content)
- Whether the same JavaScripts run multiple times (which is a requirement for some and a bug for others)
- Whether CSS is removed when it disappears from the DOM
- What happens when a new deployment changes assets on the server

Everyone has an answer for these questions in their own projects, but it differs wildly across users.

> I think that the implementation itself could be inspired by other libraries.

I have seen htmx' head merging extension. Unfortunately that approach wouldn't work for my own apps without heavy modification. I don't know Turbo's implementation.

**Reply by triskweline** [reply]
We would also need to distinguish between assets (JS, CSS) and other elements in the head.

Assets would need to be updated for every render pass, since they may be needed to compile new fragments. However, elements like `<title>`, `<link rel="canonical">`, OpenGraph tags, etc. must only be merged when we're updating history (usually when [navigating](https://unpoly.com/navigation)).

**Reply by janreges** [reply]
@triskweline - in two days HTMX will include `head-merge` extension - this is work-in-progress version: https://github.com/bigskysoftware/htmx/blob/fb4f3c48ddf93db0660c7a600a6c1d130385a49e/src/ext/head-merge.js#L85-L84 

I think that basically it contains the most essential functions.

A few days ago I developed a multi-swap extension for HTMX https://htmx.org/extensions/multi-swap/. I will also try to help expand the head-merge extension. It needs to be more configurable to handle some edge cases, but I believe we can handle it.

I will then try to project my acquired knowledge from development in HTMX into development in Unpoly. Both HTMX and Unpoly are great libraries and I see the ability to merge the <head> as really essential. Especially with the advent of access like Astro or Qwik, Unpoly can also gain popularity.

**Reply by janreges** [reply]
@triskweline - HTMX already released plugin `head-support` with documented behavior and options: https://htmx.org/extensions/head-support/

This implementation is apparently different from the one you looked at in the past - it could have come from a 2-year-old branch that was just someone's attempt.

You can check this current implementation https://github.com/bigskysoftware/htmx/blob/master/src/ext/head-support.js

Do you see any flaws in this implementation that would prevent it from working in your applications that you mentioned?

I consider this implementation to be flexible and allow solving various specific needs in projects.

**Comment by janreges** [comment]
I would like to reimplement the <head> merging functionality from HTMX to Unpoly as well. I would process it separately and then send a PR including docs update.

@triskweline - as the lead author of Unpoly and its internal architecture, could you describe to me in 2-3 sentences where in Unpoly's architecture you would put this functionality?

For now, I implement it by catching the `up:request:loaded` event and if there is an `up-head-strategy` attribute in the return <head> with the value `merge` or `append`, then I apply the required behavior, including history related requirements.

I will be very grateful for any ideas and suggestions you have on how to grasp the implementation. I believe that I will then be able to deliver a PR that will not be too much work for you :-)

**Reply by triskweline** [reply]
@janreges Thank you for offering to help with this.

I spend many evaluating possible approaches to reconcile `<head>` elements and manage the inclusion of scripts and stylesheets. The challenge here is developers using a wide variety of strategies to manage their script and styles. We need to provide a solution that works for most use cases and doesn't break existing apps.

The result is [RFC: Reconciliation of `<head>` elements](https://github.com/unpoly/unpoly/discussions/446). Let me know what if you think.

Also it's perfectly fine if you just want to port HTMX' implementation. You can do so in a plugin that hooks into the `up:fragment:loaded` event. I believe you will find everything you need in that event's properties. The plugin could not be merged into Unpoly's core, but maybe we can publish an `unpoly/unpoly-contrib` repository or link from the docs to a repo of your own.
